include/nbl/ui/IWindow.h

```diff
@@ -12,23 +16,131 @@ class IWindow : public core::IReferenceCounted
 public:
     enum E_CREATE_FLAGS : uint32_t
     {
-        ECF_FULLSCREEN = 1u<<0,
-        ECF_HIDDEN = 1u<<1,
-        ECF_BORDERLESS = 1u<<2,
-        ECF_RESIZABLE = 1u<<3,
-        ECF_MINIMIZED = 1u<<4,
-        ECF_MAXIMIZED = 1u<<5,
+        ECF_FULLSCREEN = 1u << 0,
+        ECF_HIDDEN = 1u << 1,
+        ECF_BORDERLESS = 1u << 2,
+        ECF_RESIZABLE = 1u << 3,
+        ECF_MINIMIZED = 1u << 4,
+        ECF_MAXIMIZED = 1u << 5,
         //! Forces mouse to stay inside the window
-        ECF_MOUSE_CAPTURE = 1u<<6,
+        ECF_MOUSE_CAPTURE = 1u << 6,
         //! Indicates whether the window is active or not
-        ECF_INPUT_FOCUS = 1u<<7,
+        ECF_INPUT_FOCUS = 1u << 7,
         //! Indicates whether mouse is hovering over the window even if the window is not active
-        ECF_MOUSE_FOCUS = 1u<<8,
-        ECF_ALWAYS_ON_TOP = 1u<<9,
+        ECF_MOUSE_FOCUS = 1u << 8,
+        ECF_ALWAYS_ON_TOP = 1u << 9,
 
         ECF_NONE = 0
     };
 
+
+   
+    class IEventCallback : public core::IReferenceCounted
+    {
+    public:
+        void onWindowShown(IWindow* w) 
+        {
+            w->m_flags &= (~ECF_HIDDEN);
+            onWindowShown_impl();
+        }
+        void onWindowHidden(IWindow* w)
+        {
+            w->m_flags |= ECF_HIDDEN;
+            onWindowHidden_impl();
+        }
+        void onWindowMoved(IWindow* w, int32_t x, int32_t y)
+        {
+            w->m_x = x;
+            w->m_y = y;
+            onWindowMoved_impl(x, y);
+        }
+        void onWindowResized(IWindow* w, uint32_t width, uint32_t height)
+        {
+            w->m_width = width;
+            w->m_height = height;
+            onWindowResized_impl(width, height);
+        }
+        void onWindowRotated(IWindow* w)
+        {
+            onWindowRotated_impl();
+        }
+        void onWindowMinimized(IWindow* w)
+        {
+            w->m_flags |= ECF_MINIMIZED;
+            w->m_flags &= (~ECF_MAXIMIZED);
+            onWindowMinimized_impl();
+        }
+        void onWindowMaximized(IWindow* w)
+        {
+            w->m_flags |= ECF_MAXIMIZED;
+            w->m_flags &= (~ECF_MINIMIZED);
+            onWindowMaximized_impl();
```

Comment:
state update and callback call are in reverse, also callback is supposed to let you know if it allowed the state (like resize) to happen

Reply:
@sadiuk